1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+
+`ICustomAstVisitor2` and `AstVisitor2` would be extended to deal with this AST,
+and .NET Core 3's new default interface implementation feature would be
+leveraged to ensure this does not break things
+as previous syntactic introductions have been forced to.
+
+#### Statements may not be chained
+
+PowerShell has a notable divergence from POSIX shells:
+
+- In POSIX shells, a pipeline is composed of statements
+- In PowerShell, a pipeline is a kind of statement
+
+This means that bash allows the following:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF'
+```
+
+In fact, this can be evaluated in the background:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF' &
+```
+
+However, in PowerShell, pipelines are subordinate to statements.
+The proposed implementation of chain operators precludes use of
+`&&` and `||` between statements like `if` or `while`,
+or constructions like `$x = cmd1 && $y = cmd2 && $x + $y`.
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+## Related Material
+
+- Original issue: [PowerShell/PowerShell #3241](https://github.com/PowerShell/PowerShell/issues/3241).
+
+- Work-in-progress implementation: [PowerShell/PowerShell #9849](https://github.com/PowerShell/PowerShell/pull/9849).
+
+- [Current handling of `&&` and `||` in the parser](https://github.com/PowerShell/PowerShell/blob/af1de9e88d28014438ff3414e82298e5b14f6e81/src/System.Management.Automation/engine/parser/Parser.cs#L5846-L5860).
+
+## Alternate Proposals and Considerations
+
+### `&&` and `||` as a statement separator
+
+To be more bash-like, an alternative implementation might treat
+`&&` and `||` with the same precedence as `;`, as a way to separate statements.
+
+This would allow assignment within chains, like:
+
+```powershell
+$x = cmd1 && $y = cmd2 && $x + $y
+```
+
+Other possibilities would include:
+
+```powershell
+if ($condition) { cmd1 } && while ($anotherCondition) { cmd2 }
+```
+
+```powershell
+foreach ($v in 1..100)
+{
+    cmd1 $v && break
+}
+```
+
+#### Background operator changes
+
+In bash, background operators are lower precedence than chain separators
+(`&` has the same precedence as `;`).
+In PowerShell, background operators are higher precedence.
+
+Making chain operators apply to any statement would lead to the question
+of the precedence of the background operator.
+
+The example:
+
+```powershell
+cmd1 && cmd2 &
+```
+
+Could either be `[[cmd1 && cmd2] &]` or `[cmd1 && [cmd2 &]]`.
+
+Having chain operators apply to statements would mean the second case
+is the correct one, unless changes are made to the background operator precedence (which would also allow `if ($condition) { cmd1 } &)` for example).
+
+The use of this is possibly less than the use of being able to background
+an entire pipeline chain; `cmd1 & && cmd2` would be equivalent to
+`cmd1 &; cmd2` since a backgrounded pipeline
+will never fail in the invoking context.
+

```

Comment:
If you want to background an entire pipeline chain, wrap it in a script block invoked with the call operator and use the background operator with that script block (once [this bug](https://github.com/PowerShell/PowerShell/issues/9703) is fixed for the background operator).

Speaking of script blocks, this RFC doesn't talk about them, but bash supports commands like this:

`command1 && {command2; command3;} || command4`

We should consider doing the same.

Reply:
so `command1 && {command2; command3;} || command4` would actually be equivalent to `command1 && $(& { command2; command3 }) || command4`? I'm not sure I like that implied invocation. What if I want that scriptblock to be sent to output or captured as an RVALUE?

Reply:
Let's use `&&&` for that! I'm kidding -- I think `command1 && $(command2; command3) || command4` is sufficient and not worth the syntactic sugar of curly braces (nor the added ambiguity.)

Reply:
Yeah `cmd1 && $(cmd2; cmd3) && cmd4` is a valid syntax under the proposal (`$(...)` is a valid pipeline), but unfortunately always sets `$?` to true.

I think the answer to that is to fix `$?` though.

Reply:
Make it so? :)

Reply:
Subexpressions really don't feel like the right answer here.

My argument for simply dot-sourcing a script block was because that is what users would want the vast majority of the time from that syntax, and for the other times when they do indeed want a script block they can still get one. That said, I recognize that it diverges from the current syntax.

I feel subexpressions are really the wrong syntax to promote for invocation of multiple commands in a staged execution statement like this though, and I worry that's likely to lead to confusion and users trying to use them in other locations that they cannot or in situations where script blocks are a better answer.

@rjmholt: The following should work, correct?

```powershell
pipeline1 && . {command1; command2} || pipeline3
```

As long as that works, I don't have much issue with this.

Reply:
> The following should work, correct?

Yeah that will work perfectly.

Without the `.`, it will also execute successfully but you'll be passing a scriptblock expression into the pipeline (i.e. it won't do what's intended)